routes_handler.go

```diff
@@ -18,93 +17,159 @@ import (
-			} else if strings.Contains(err.Error(), "unknown chain hash") {
-				http.Error(w, "unknown chain hash", http.StatusBadRequest)
+			slog.Error("Failed get beacon", "error", err, "nextTime", nextTime)
+
+			if nextTime < 0 {
+				w.Header().Set("Cache-Control", fmt.Sprintf("must-revalidate, public, max-age=%d", -nextTime))
+
+				// I know, 425 is meant to indicate a replay attack risk, but hey, it's the perfect error name!
+				http.Error(w, "Requested future beacon", http.StatusTooEarly)
 			} else {
+				w.Header().Set("Cache-Control", "must-revalidate, no-cache, max-age=0")
+
 				http.Error(w, "Failed to get beacon", http.StatusInternalServerError)
 			}
 			return
 		}
 
-		nextTime, nextRound := info.ExpectedNext()
-		if round >= nextRound+1 { // never happens when fetching latest because round == 0
-			w.Header().Set("Cache-Control", "must-revalidate, no-cache, max-age=0")
-			slog.Error("[GetBeacon] Future beacon was requested, unexpected", "requested", round, "expected", nextRound, "from", r.RemoteAddr)
-			// I know, 425 is meant to indicate a replay attack risk, but hey, it's the perfect error name!
-			http.Error(w, "Requested future beacon", http.StatusTooEarly)
-			return
-		} else if round == nextRound {
-			// we wait until the round is supposed to be emitted, minus frontrun to account for network latency anyway
-			time.Sleep(time.Duration(nextTime-time.Now().Unix())*time.Second - FrontrunTiming)
-		}
+		writeBeacon(w, beacon, nextTime, isV2)
+	}
+}
+
+func getBeacon(c *grpc.Client, r *http.Request, round uint64) (*grpc.HexBeacon, int64, error) {
+	m, err := createRequestMD(r)
+	if err != nil {
+		return nil, 0, fmt.Errorf("createRequestMD error: %w", err)
+	}
 
-		beacon, err := c.GetBeacon(r.Context(), m, round)
+	info, err := c.GetChainInfo(r.Context(), m)
+	if err != nil {
+		return nil, 0, fmt.Errorf("GetChainInfo error: %w", err)
+	}
+
+	nextTime, nextRound := info.ExpectedNext()
+	// we refuse rounds too far in the future
+	if round >= nextRound+1 {
+		slog.Error("[GetBeacon] Future beacon was requested, unexpected", "requested", round, "expected", nextRound, "from", r.RemoteAddr)
+		// TODO: we could have a more precise nextTime value instead of just period
+		// we return the negative time to cache this response
+		return nil, -int64(info.Period), fmt.Errorf("future beacon was requested")
+	}
+
+	var beacon *grpc.HexBeacon
+	// if we are requesting the next round
+	if round == nextRound {
+		beacon, err = c.Next(r.Context(), m)
 		if err != nil {
-			if err != nil {
-				slog.Error("all clients are unable to provide beacons", "error", err)
-				w.Header().Set("Cache-Control", "must-revalidate, no-cache, max-age=0")
-				http.Error(w, "Failed to get beacon", http.StatusInternalServerError)
-				return
-			}
+			slog.Error("[GetBeacon] unable to get next beacon from any grpc client", "error", err)
+			return nil, 0, fmt.Errorf("Next error: %w", err)
+
+		}
+		// we use -1 to indicate no caching
+		nextTime = -1
```

Comment:
doesn't this result in the error being cached for 1 second on L37? https://github.com/drand/http-relay/pull/10/files#diff-d15942135d82d9e657d60b16d725e6f38cef3994c889f58e4fd8d3abc2e32ab7R37